pkg/resourcegroup/controllers/root/root_controller_test.go

```diff
@@ -139,63 +156,82 @@ func TestRootReconciler(t *testing.T) {
 	}
-
-	// The watchmap should be updated correctly
-	for _, r := range []*Reconciler{reconcilerKpt} {
-		watched := r.watches.IsWatched(schema.GroupVersionKind{
-			Group: "apps", Version: "v1", Kind: "Deployment"})
-		assert.True(t, watched)
-		watched = r.watches.IsWatched(schema.GroupVersionKind{
-			Group: "apps", Version: "v1", Kind: "StatefulSet"})
-		assert.True(t, watched)
-		watched = r.watches.IsWatched(schema.GroupVersionKind{
-			Group: "apps", Version: "v1", Kind: "DaemonSet"})
-		assert.True(t, watched)
-		assert.Equal(t, 3, r.watches.Len())
+	if t.Failed() {
+		t.FailNow()
 	}
 
-	// There should be one event pushed to the channel.
-	go func() { e = <-reconcilerKpt.channel }()
-	time.Sleep(time.Second)
-	assert.Equal(t, "group0", e.Object.GetName())
-
-	// Delete the resource group
+	// Wait for the TypeResolver to update the watches
+	require.EventuallyWithT(t, func(t *assert.CollectT) {
+		assert.Equal(t, 3, reconcilerKpt.watches.Len())
+		for _, r := range []*Reconciler{reconcilerKpt} {
+			watched := r.watches.IsWatched(schema.GroupVersionKind{
+				Group: "apps", Version: "v1", Kind: "Deployment"})
+			assert.True(t, watched)
+			watched = r.watches.IsWatched(schema.GroupVersionKind{
+				Group: "apps", Version: "v1", Kind: "StatefulSet"})
+			assert.True(t, watched)
+			watched = r.watches.IsWatched(schema.GroupVersionKind{
+				Group: "apps", Version: "v1", Kind: "DaemonSet"})
+			assert.True(t, watched)
+			assert.Equal(t, 3, r.watches.Len())
+		}
+	}, 5*time.Second, time.Second)
+
+	t.Log("Deleting ResourceGroup")
 	err = c.Delete(ctx, resourceGroupKpt)
-	assert.NoError(t, err)
+	require.NoError(t, err)
+
+	// Wait for the TypeResolver to update the watches
+	require.EventuallyWithT(t, func(t *assert.CollectT) {
+		assert.Equal(t, 0, reconcilerKpt.watches.Len())
+	}, 5*time.Second, time.Second)
+
+	// Validate that the Root controller didn't send a channel event
+	select {
+	case e := <-reconcilerKpt.channel:
+		t.Fatalf("expected channel NOT to be sent an event, but got %+v", e)
+	default: // success
+	}
 
-	// The delete triggers another reconcile
-	// wait until it ends.
-	time.Sleep(2 * time.Second)
+	// Validate that the Root controller updated the resMap
 	assert.NotNil(t, reconcilerKpt.resMap)
-
-	// The resmap should be updated correctly
-	// It doesn't contain any resourcegroup or resource
 	assert.False(t, reconcilerKpt.resMap.HasResgroup(request.NamespacedName))
-	assert.NotNil(t, reconcilerKpt.resMap)
 	for _, resource := range resources {
 		assert.False(t, reconcilerKpt.resMap.HasResource(resource))
 	}
+}
 
-	// There should be one event pushed to the channel.
-	go func() { e = <-reconcilerKpt.channel }()
-	time.Sleep(time.Second)
-	assert.Equal(t, "group0", e.Object.GetName())
+func waitForEvent(t *testing.T, inCh <-chan event.GenericEvent, timeout time.Duration) {
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+	select {
+	case e := <-inCh:
+		require.Equal(t, "group0", e.Object.GetName())
```

Comment:
If the name (`group0`) is not passed as a parameter, suggest moving it to a constant to make it clear that it's assumed this is the only `ResourceGroup` name used for tests

Reply:
Done.